File: 백준/Silver/9934. 완전 이진 트리/완전 이진 트리.py
import sys
from collections import deque

k = int(input())
k = 2**k - 1
arr = list(map(int, sys.stdin.readline().split()))

class Node:
    def __init__(self, val = 0, left = None, right = None):
        self.val = val
        self.left = left
        self.right = right
        
class BinaryTree:
    def __init__(self):
        self.root = None
        self.idx = 0
    
    def add_node(self, k):
        if k == 0:
            return
        
        self.root = Node()
        q = deque([self.root])
        i = 1
        
        while i < k:
            cur_node = q.popleft()
            
            cur_node.left = Node()
            q.append(cur_node.left)
            i += 1
            
            cur_node.right = Node()
            q.append(cur_node.right)
            i += 1
        
    def add_value(self, cur_node):
        if cur_node == None:
            return
        
        left = self.add_value(cur_node.left)
        
        if left == None:
            cur_node.val = arr[self.idx]
            self.idx += 1

        self.add_value(cur_node.right)
    
    def print_level_order(self, cur_node):
        visited = []
        if cur_node == None:
            return
        
        q = deque()
        q.append((cur_node, 0))
        
        while q:
            cur_node, cur_depth = q.popleft()
            
            if len(visited) <= cur_depth:
                visited.append([])
            
            visited[cur_depth].append(cur_node.val)

            if cur_node.left:
                q.append((cur_node.left, cur_depth + 1))
            
            if cur_node.right:
                q.append((cur_node.right, cur_depth + 1))
        
        for i in range(len(visited)):
            for j in range(len(visited[i])):
                print(visited[i][j], end=' ')
            print()
            
bt = BinaryTree()
bt.add_node(k)
bt.add_value(bt.root)
bt.print_level_order(bt.root)

File: 백준/Silver/9934. 완전 이진 트리/완전 이진 트리.java
import com.sun.source.tree.Tree;

import java.io.*;
import java.util.*;

public class Main {
    static int[] nums;
    static int idx = 0;
    static StringBuilder sb = new StringBuilder();

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        int k = Integer.parseInt(br.readLine());
        k = (int) Math.pow(2, k) - 1;
        nums = new int[k];
        st = new StringTokenizer(br.readLine());

        for(int i = 0; i < nums.length; i++) nums[i] = Integer.parseInt(st.nextToken());

        TreeNode root = add_node(k);
        add_value(root);
        level_order(root);

        System.out.println(sb);
        br.close();
    }

    static void level_order(TreeNode root) {
        Queue<Nodes> q = new LinkedList<>();
        q.offer(new Nodes(root, 0));
        int depth = 0;

        while (!q.isEmpty()) {
            Nodes cur_root = q.poll();
            TreeNode cur_node = cur_root.root;
            int cur_depth = cur_root.depth;

            if (depth != cur_depth) {
                sb.append("\n");
                depth = cur_depth;
            }

            sb.append(cur_node.value).append(" ");

            if (cur_node.left != null) q.offer(new Nodes(cur_node.left, cur_depth + 1));
            if (cur_node.right != null) q.offer(new Nodes(cur_node.right, cur_depth + 1));
        }
    }

    static TreeNode add_node(int k) {
        Queue<TreeNode> q = new LinkedList<>();
        TreeNode root = new TreeNode();
        q.offer(root);

        while ((k -= 2) > 0) {
            TreeNode cur_node = q.poll();

            cur_node.left = new TreeNode();
            q.offer(cur_node.left);

            cur_node.right = new TreeNode();
            q.offer(cur_node.right);
        }

        return root;
    }

    static void add_value(TreeNode cur_node) {
        if (cur_node == null) return;

        add_value(cur_node.left);

        cur_node.value = nums[idx];
        idx++;

        add_value(cur_node.right);
    }

    static class Nodes {
        TreeNode root;
        int depth;

        Nodes(TreeNode root, int depth) {
            this.root = root;
            this.depth = depth;
        }
    }

    static class TreeNode {
        int value;
        TreeNode left;
        TreeNode right;

        TreeNode() {
            this.value = 0;
            this.left = null;
            this.right = null;
        }
    }
}